common/persistence/elasticsearch/esProcessor.go

```diff
@@ -247,81 +283,96 @@ func (p *esProcessorImpl) sendToAckChan(visibilityTaskKey string, ack bool) {
 	})
 }
 
-func (p *esProcessorImpl) getVisibilityTaskKey(request elastic.BulkableRequest) string {
+func (p *esProcessorImpl) extractVisibilityTaskKey(request elastic.BulkableRequest) string {
 	req, err := request.Source()
 	if err != nil {
-		p.logger.Error("Get request source err.", tag.Error(err), tag.ESRequest(request.String()))
+		p.logger.Error("Unable to get ES request source.", tag.Error(err), tag.ESRequest(request.String()))
 		p.metricsClient.IncCounter(metrics.ElasticSearchVisibility, metrics.ESBulkProcessorCorruptedData)
 		return ""
 	}
 
-	var key string
 	if len(req) == 2 { // index or update requests
 		var body map[string]interface{}
-		if err := json.Unmarshal([]byte(req[1]), &body); err != nil {
-			p.logger.Error("Unmarshal index request body err.", tag.Error(err))
+		if err = json.Unmarshal([]byte(req[1]), &body); err != nil {
+			p.logger.Error("Unable to unmarshal ES request body.", tag.Error(err))
 			p.metricsClient.IncCounter(metrics.ElasticSearchVisibility, metrics.ESBulkProcessorCorruptedData)
 			return ""
 		}
 
 		k, ok := body[definition.VisibilityTaskKey]
 		if !ok {
-			// must be bug in code and bad deployment, check processor that add es requests
-			panic("VisibilityTaskKey not found")
-		}
-		key, ok = k.(string)
-		if !ok {
-			// must be bug in code and bad deployment, check processor that add es requests
-			panic("VisibilityTaskKey is not string")
-		}
-	} else { // delete requests
-		var body map[string]map[string]interface{}
-		if err := json.Unmarshal([]byte(req[0]), &body); err != nil {
-			p.logger.Error("Unmarshal delete request body err.", tag.Error(err))
+			p.logger.Error("Unable to extract VisibilityTaskKey from ES request.", tag.ESRequest(request.String()))
 			p.metricsClient.IncCounter(metrics.ElasticSearchVisibility, metrics.ESBulkProcessorCorruptedData)
 			return ""
 		}
+		return k.(string)
+	} else { // delete requests
+		return p.extractDocID(request)
+	}
+}
 
-		opMap, ok := body["delete"]
-		if !ok {
-			// must be bug, check if dependency changed
-			panic("delete key not found in request")
-		}
-		k, ok := opMap["_id"]
-		if !ok {
-			// must be bug in code and bad deployment, check processor that add es requests
-			panic("_id not found in request opMap")
+func (p *esProcessorImpl) extractDocID(request elastic.BulkableRequest) string {
```

Comment:
[Nit] just have this return (string, error) instead of relying on "" to signal issue?

Reply:
`_id` must always be there.